Allow mixing debug and release builds

msimberg opened the pull request:
Fixes #5013.

Testing if this is feasible. In principle it is, but I'd appreciate feedback on if there are any pitfalls that I'm not aware of, especially on Windows. Ideally we only use `HPX_DEBUG` determine if parts of the code should be conditionally compiled/disabled/etc. However, there are a few additional macros that we use:

- `DEBUG`: I could find no reference to anything using this, does anyone know what it's doing in HPX?
- `_DEBUG`: Windows specific, added automatically in debug mode, correct? There are a couple of uses of this one. Can we replace them with `HPX_DEBUG`? (https://github.com/STEllAR-GROUP/hpx/blob/8cb1829d7ae94d1cc078a7ed1a6014c212406a09/src/runtime_distributed.cpp#L546-L551)
- `NDEBUG`: Disables `assert`. We have a few uses of `assert` which are fine, but as far as I can tell there is no reason for us to rely on this macro.
- `HPX_DISABLE_ASSERTIONS`/`BOOST_DISABLE_ASSERTIONS`: Used to disable a few assertions, but not `HPX_ASSERT`. Used together with `NDEBUG` in some cases. I'd at most leave `HPX_DISABLE_ASSERTIONS`.

Is there a use case for debug builds with assertions disabled? If yes, should it get a CMake option instead of needing to be manually added to the compile flags?

hkaiser commented:
> * `DEBUG`: I could find no reference to anything using this, does anyone know what it's doing in HPX?

There is no reason to use this anywhere, even more as we map it onto `HPX_DEBUG` (https://github.com/STEllAR-GROUP/hpx/blob/master/libs/core/config/include/hpx/config/debug.hpp#L19-L25)

> * `_DEBUG`: Windows specific, added automatically in debug mode, correct? 

Yes, this auto-added in Debug builds. It is mapped onto `HPX_DEBUG`, also here: https://github.com/STEllAR-GROUP/hpx/blob/master/libs/core/config/include/hpx/config/debug.hpp#L19-L25

> There are a couple of uses of this one. Can we replace them with `HPX_DEBUG`? 

Yes.

> * `NDEBUG`: Disables `assert`. We have a few uses of `assert` which are fine, but as far as I can tell there is no reason for us to rely on this macro.

This is auto-added on windows for non-debug builds, there shouldn't be any reason for us to use it. Also, we should replace all `assert()` with `HPX_ASSERT()`. Actually, I thought we have an inspect check for this.

> * `HPX_DISABLE_ASSERTIONS`/`BOOST_DISABLE_ASSERTIONS`: Used to disable a few assertions, but not `HPX_ASSERT`. Used together with `NDEBUG` in some cases. I'd at most leave `HPX_DISABLE_ASSERTIONS`.
> 
> Is there a use case for debug builds with assertions disabled? If yes, should it get a CMake option instead of needing to be manually added to the compile flags?

I think this is a leftover from times when this macro caused `HPX_ASSERT` to fallback to `assert`. I think this can be removed altogether.

msimberg commented:
> > * `DEBUG`: I could find no reference to anything using this, does anyone know what it's doing in HPX?
> 
> There is no reason to use this anywhere, even more as we map it onto `HPX_DEBUG` (https://github.com/STEllAR-GROUP/hpx/blob/master/libs/core/config/include/hpx/config/debug.hpp#L19-L25)
> 
> > * `_DEBUG`: Windows specific, added automatically in debug mode, correct?
> 
> Yes, this auto-added in Debug builds. It is mapped onto `HPX_DEBUG`, also here: https://github.com/STEllAR-GROUP/hpx/blob/master/libs/core/config/include/hpx/config/debug.hpp#L19-L25
> 
> > There are a couple of uses of this one. Can we replace them with `HPX_DEBUG`?
> 
> Yes.
> 
> > * `NDEBUG`: Disables `assert`. We have a few uses of `assert` which are fine, but as far as I can tell there is no reason for us to rely on this macro.
> 
> This is auto-added on windows for non-debug builds, there shouldn't be any reason for us to use it. Also, we should replace all `assert()` with `HPX_ASSERT()`. Actually, I thought we have an inspect check for this.
> 
> > * `HPX_DISABLE_ASSERTIONS`/`BOOST_DISABLE_ASSERTIONS`: Used to disable a few assertions, but not `HPX_ASSERT`. Used together with `NDEBUG` in some cases. I'd at most leave `HPX_DISABLE_ASSERTIONS`.
> > 
> > Is there a use case for debug builds with assertions disabled? If yes, should it get a CMake option instead of needing to be manually added to the compile flags?
> 
> I think this is a leftover from times when this macro caused `HPX_ASSERT` to fallback to `assert`. I think this can be removed altogether.

Perfect, thanks @hkaiser for confirming all these!

> > * `_DEBUG`: Windows specific, added automatically in debug mode, correct?
> 
> Yes, this auto-added in Debug builds. It is mapped onto `HPX_DEBUG`, also here: https://github.com/STEllAR-GROUP/hpx/blob/master/libs/core/config/include/hpx/config/debug.hpp#L19-L25

I was hoping I could remove that mapping (it's already gone on this branch), since with the mapping we can still end up with inconsistent flags (user code code compiled with `HPX_DEBUG` defined and HPX without it).

msimberg commented:
> > * `NDEBUG`: Disables `assert`. We have a few uses of `assert` which are fine, but as far as I can tell there is no reason for us to rely on this macro.
> 
> This is auto-added on windows for non-debug builds, there shouldn't be any reason for us to use it. Also, we should replace all `assert()` with `HPX_ASSERT()`. Actually, I thought we have an inspect check for this.

This is not as bad as it sounds. I rechecked and we have it in the moodycamel queue header. It's already excluded from inspect. It might be worth leaving the `assert`s for easier merging of new versions, but then again it already has formatting and a few other changes to make it different so we could also change them. Don't really mind.

The other place where we use `assert` is for assertions in device code, which I also think is fine.

hkaiser commented on cmake/HPX_CompilerFlagsTargets.cmake:
```diff
@@ -21,6 +21,13 @@ target_compile_definitions(
+target_compile_definitions(
+  hpx_private_flags
+  INTERFACE $<$<CONFIG:MinSizeRel>:NDEBUG> hpx_private_flags
+  INTERFACE $<$<CONFIG:Release>:NDEBUG> hpx_private_flags
```
Do we need the private flags twice?

msimberg replied:
That's a bug, it shouldn't be there at all. Good catch!

hkaiser commented on cmake/templates/HPXMacros.cmake.in:
```diff
@@ -95,25 +95,3 @@ function(hpx_check_allocator_compatibility)
-        "debug modes).")
-    endif()
-  endif()
-endfunction()
```
Should we at least leave a warning?

msimberg replied:
I'm not sure. My thinking was that this is now fully the users responsibility and mixing build types is a normal thing to do, but I can add back a warning if you feel it's appropriate.

hkaiser replied:
We don't need a warning if we can ensure that everything works without issues.

hkaiser reviewed (commented):
Should we add CI configurations making sure using a Debug version from a release build and v.v. is actually functional? I'd like to see that for all major platforms (at least Linux and Windows).

msimberg commented:
> Should we add CI configurations making sure using a Debug version from a release build and v.v. is actually functional? I'd like to see that for all major platforms (at least Linux and Windows).

That's my last concern. The build unit tests now build the hello world component example in four different build types. It's not exhaustive but it's a start. A more comprehensive test would be to make use of the feature to build tests against an installed HPX. Let's at the very least wait until we can get the windows builders running again.

msimberg commented:
retest lsu

hkaiser commented:
This issue is not directly connected, but it's related: https://github.com/STEllAR-GROUP/phylanx/issues/1311

msimberg commented on cmake/templates/HPXMacros.cmake.in:
```diff
@@ -95,25 +95,3 @@ function(hpx_check_allocator_compatibility)
-function(hpx_check_cmake_build_type)
-  get_property(_GENERATOR_IS_MULTI_CONFIG GLOBAL PROPERTY GENERATOR_IS_MULTI_CONFIG)
-  if (NOT _GENERATOR_IS_MULTI_CONFIG AND
-      NOT "${HPX_BUILD_TYPE}" STREQUAL "${CMAKE_BUILD_TYPE}" AND
```
@hkaiser, @K-ballo, we used to skip checking that build types match for Visual Studio projects anyway. Is there anything we need to or can do to check that, or is that handled somehow automatically by Visual Studio?

hkaiser replied:
@msimberg I'm not sure we can do anything about it at configuration time as the generated VS projects are multi-target. The mismatch should be caught by the versioning symbols we generate, though (when building applications).

msimberg commented:
@hkaiser any idea what I might've missed here: https://github.com/STEllAR-GROUP/hpx/pull/5099/checks?check_run_id=1870902759#step:5:1476?

hkaiser commented:
> @hkaiser any idea what I might've missed here: https://github.com/STEllAR-GROUP/hpx/pull/5099/checks?check_run_id=1870902759#step:5:1476?

Let me try to reproduce this locally.

hkaiser commented:
> > @hkaiser any idea what I might've missed here: https://github.com/STEllAR-GROUP/hpx/pull/5099/checks?check_run_id=1870902759#step:5:1476?
> 
> Let me try to reproduce this locally.

Looks like that the same commands just work fine for me locally. So it has to be some resource restriction imposed by github. We could try to use `-verbosity:detailed` (instead of the `-verbosity:minimal`) and hope to see more information.

msimberg commented:
> > > @hkaiser any idea what I might've missed here: https://github.com/STEllAR-GROUP/hpx/pull/5099/checks?check_run_id=1870902759#step:5:1476?
> > 
> > 
> > Let me try to reproduce this locally.
> 
> Looks like that the same commands just work fine for me locally. So it has to be some resource restriction imposed by github. We could try to use `-verbosity:detailed` (instead of the `-verbosity:minimal`) and hope to see more information.

Thanks for checking. I'll try that out.

hkaiser commented:
@msimberg I pushed some changes to this branch that work around an MSVC issue that suddenly manifested itself on this branch. Let's see if that fixes the problems for the Windows builders,

hkaiser commented:
FWIW, I think the remaining issue is caused by the additional targets you added (`cmake_debug_build_dir_targets_test` and friends).

msimberg commented on .github/workflows/windows_release.yml:
```diff
@@ -38,19 +38,28 @@ jobs:
+              cmake_release_install_dir_targets_test.make_compile   \
+              cmake_release_install_dir_macros_test.make_build_dir \
+              cmake_release_install_dir_macros_test.make_configure \
+              cmake_release_install_dir_macros_test.make_compile   \
```
Grr, sorry @hkaiser for wasting your time with this. Thanks for catching this!

hkaiser commented:
@msimberg this looks like it's good to go.

msimberg commented:
retest lsu

msimberg commented:
> @msimberg this looks like it's good to go.

Thanks you @hkaiser for helping with this!

rbyshko commented:
I have installed hpx 1.6.0 using vcpkg and tried to build a "Hello world" program as in Quick Start from the documentation. Unfortunately, I still get the error this issue claims to have fixed.

With vcpkg I have also installed the current master (4d748091cc937b9eb61b7b42562191e1e2b638e3), but the error is still there.

```
terminate called after throwing an instance of 'hpx::detail::exception_with_info<hpx::exception>'
  what():  failed to insert base_set_event_action into typename to id registry.: HPX(invalid_status)

```
Let me know if I can provide more information to get this fixed.

msimberg commented:
Since you mention vcpkg I'm guessing you're on Windows, is that correct? If that's the case this PR does not allow mixing build types on Windows. I don't know if we can do more to detect these situations, any help with that would definitely be appreciated.

rbyshko commented:
No, I am using vcpkg on Linux, Ubuntu 20.04 to be exact. vcpkg installs release and debug versions of the hpx. However they go into separate directories. CMake seems to handle it correctly:

![image](https://user-images.githubusercontent.com/35375029/110829353-4c334380-8298-11eb-916e-ae651bfbef9e.png)

msimberg commented:
> No, I am using vcpkg on Linux, Ubuntu 20.04 to be exact. vcpkg installs release and debug versions of the hpx. However they go into separate directories. CMake seems to handle it correctly:
> 
> ![image](https://user-images.githubusercontent.com/35375029/110829353-4c334380-8298-11eb-916e-ae651bfbef9e.png)

Ok, thanks for the details! That does sound like something I've missed in that case, I'll try to have a look at this soon. As you can see this is rather new, and our testing (clearly) isn't thorough enough for it.

hkaiser commented:
> I have installed hpx 1.6.0 using vcpkg and tried to build a "Hello world" program as in Quick Start from the documentation. Unfortunately, I still get the error this issue claims to have fixed.
> 
> With vcpkg I have also installed the current master ([4d74809](https://github.com/STEllAR-GROUP/hpx/commit/4d748091cc937b9eb61b7b42562191e1e2b638e3)), but the error is still there.
> 
> ```
> terminate called after throwing an instance of 'hpx::detail::exception_with_info<hpx::exception>'
>   what():  failed to insert base_set_event_action into typename to id registry.: HPX(invalid_status)
> ```
> 
> Let me know if I can provide more information to get this fixed.

This error message indicates that both versions, release and debug, have been loaded. Apparently we still miss something as it shouldn't actually happen. The plugin API we use relies on different function names for release and debug, so there is at least a potential way of preventing to load a mismatched version.

msimberg commented:
@rbyshko just for completeness, could you provide the code and CMakeLists.txt that you used? And are you seeing this problem with user code in release mode and HPX in debug mode, vice versa, or both?

rbyshko commented:
I am experiencing this error in both release and debug builds.

CMakeLists.txt
```
cmake_minimum_required(VERSION 3.16)

set(CMAKE_TOOLCHAIN_FILE ${CMAKE_CURRENT_LIST_DIR}/vcpkg/scripts/buildsystems/vcpkg.cmake)

project(HPX)

find_package(HPX CONFIG REQUIRED)

add_executable(hpx_hello hpx_hello.cpp)
target_link_libraries(hpx_hello PRIVATE HPX::hpx HPX::wrap_main HPX::iostreams_component)
```

hpx_hello.cpp
```
#include <hpx/hpx_main.hpp>
#include <hpx/iostream.hpp>

int main()
{
    // Say hello to the world!
    hpx::cout << "Hello World!\n" << hpx::flush;
    return 0;
}

```
